### ISSUE TITLE
Network example produces different results on Linux vs Windows

### ISSUE THREAD

**Issue by vguhesan** [NFSdb/nfsdb#28]
I ran the SimpleReplication server/client example as outlined in this URL:
http://github.nfsdb.org/

The key difference between the example shown on the website and mine is that I declare a serverConfig or clientConfig with a setIdName("eth0"):

[server]
JournalServer server = new JournalServer(new ServerConfig() {{
            setIfName("eth0");
        }}, factory);

[client]
final JournalClient client = new JournalClient(new ClientConfig() {{
            setIfName("eth0");
        }},factory);

Expected Behavior:
1. As soon as I declare a JournalReader on the client, the db file must be immediately synced.
2. As the server publishes new data, it should be synchronized on the client side.

On Windows (as expected), as soon as I run the SimpleReplicationClientMain, I can see the sync messages as shown below:

C:\VG\dev\NFSDBTest\src>java SimpleReplicationClientMain C:\\VG\dev\\NFSDBTest\\data
Jan 05, 2015 4:51:42 PM com.nfsdb.journal.net.mcast.AbstractOnDemandPoller
INFO: Polling on name:eth0 (Intel(R) 82579LM Gigabit Network Connection)
Jan 05, 2015 4:51:42 PM com.nfsdb.journal.net.config.ClientConfig
INFO: Connected to /192.168.1.12:7075
Client started
took: 1978, count=1000000
Jan 05, 2015 4:51:42 PM com.nfsdb.journal.net.StatsCollectingReadableByteChannel

INFO: received 39000097 bytes @ 153.059222 MB/s from: /192.168.1.12:7075 [1939 c
alls]
took: 406, count=1000000

Whereas running the same on a Linux machine, you can see that the client connected fine but no sync messages for a long time. And then about 15-min later, you start seeing the sync messages but the count remains zero. The debug messages keeps repeating for a minute and then pauses for another minute and this keeps repeating in a cycle.

DEBUG on {LINUX} SERVER:
INFO: sent 71 bytes @ Infinity MB/s to: /192.168.1.19:43112 [4 calls]
Jan 05, 2015 7:55:51 PM com.nfsdb.journal.net.StatsCollectingWritableByteChannel

DEBUG on {LINUX} CLIENT:
took: 1420505748619, count=0
Jan 05, 2015 7:55:48 PM com.nfsdb.journal.net.StatsCollectingReadableByteChannel
INFO: received 57 bytes @ 0.001394 MB/s from: /192.168.1.19:7075 [3 calls]

*** Version Details ***
nfsdb-core-2.1.1-SNAPSHOT.jar
disruptor-3.3.0.jar

$ lsb_release -a
No LSB modules are available.
Distributor ID:    Ubuntu
Description:    Ubuntu 14.04.1 LTS
Release:    14.04
Codename:    trusty

$ java -version
java version "1.8.0_25"
Java(TM) SE Runtime Environment (build 1.8.0_25-b17)
Java HotSpot(TM) 64-Bit Server VM (build 25.25-b02, mixed mode)

*** Noticed the same behavior on Ubuntu running OpenJDK *** 

Please advise. I can try this on a CentOS system as well in the next few days and publish my results.

**Comment by bluestreak01** [comment]
I cannot think of an obvious reason why this shouldn't be working. I'm going to try to reproduce this and come back asap.

**Comment by bluestreak01** [comment]
I suspect this problem is to do with both client and server running on the same machine and using external interface. I would assume that network driver has to be smart enough to realise that this is "loopback" setup, which it may or may not do. Sure way to test it out is to use separate machines or explicit loopback interface.

I suspect Ubuntu driver isn't as "smart" as windows one.

I'll try to test this theory but this behaviour could be specific to network driver you are using.

**Comment by vguhesan** [comment]
Vlad, 
it would be great to figure out why this is the case and document a work around (if such is the case). 

Would setting both the server and client (config) with setHostname("localhost") do the trick? In my /etc/hosts file, I have localhost defined as 127.0.0.1. Please advise. 

I can try this out tonight and let you know my results. Also if you're unable to reproduce this issue on your side, I am in the EST timezone so if you need me to setup a Join.me session, I can do that as well. It will have to be in the evenings in EST. Let me know.

**Comment by bluestreak01** [comment]
Using localhost will most likely work fine. i should be able to try Ubuntu tonight as I just realised a quick way to do it. Might be easier than wasting your time!

If you are looking for same box replication there is a way to do it without network. I'll give you a code example if you like?

**Comment by vguhesan** [comment]
Just tried the the following for both the serverConfig and clientConfig. Same results. No sync.

**** 1st test ****
setIfName("eth0");
setHostname("localhost");
****

[Server]
$ ./server-run.sh 
Jan 06, 2015 4:48:24 PM com.nfsdb.journal.net.config.ServerConfig
INFO: Server is now listening on localhost/127.0.0.1:7075 [eth0]
Publishing: Jan 06, 2015 4:48:24 PM com.nfsdb.journal.net.mcast.AbstractOnDemandSender
INFO: Sending on: name:eth0 (eth0)
Jan 06, 2015 4:48:26 PM com.nfsdb.journal.net.JournalServer
INFO: Connected: /127.0.0.1:43386
.......... [Done]

[Client]
$ ./client-run.sh 
Jan 06, 2015 4:48:26 PM com.nfsdb.journal.net.config.ClientConfig
INFO: Connected to localhost/127.0.0.1:7075
Client started

*****
[server-run.sh]
#!/bin/bash
java -classpath "/datastore/NFSDBTest/lib/disruptor-3.3.0.jar:/datastore/NFSDBTest/lib/nfsdb-core-2.1.1-SNAPSHOT.jar:/datastore/NFSDBTest/src/" SimpleReplicationServerMain /datastore/NFSDBTest/data

[client-run.sh]
#!/bin/bash
java -classpath "/datastore/NFSDBTest/lib/disruptor-3.3.0.jar:/datastore/NFSDBTest/lib/nfsdb-core-2.1.1-SNAPSHOT.jar:/datastore/NFSDBTest/src/" SimpleReplicationClientMain /datastore/NFSDBTest/data2

*****
**** 2nd test ****
setHostname("localhost");
setEnableMulticast(false);  
#Needed to set Multicast to false since multicast was not allowed in the "lo" interface
****
$ ./server-run.sh 
Jan 06, 2015 5:03:11 PM com.nfsdb.journal.net.config.ServerConfig
INFO: Server is now listening on localhost/127.0.0.1:7075 [lo]
Publishing: .Jan 06, 2015 5:03:14 PM com.nfsdb.journal.net.JournalServer
INFO: Connected: /127.0.0.1:43474
......... [Done]

$ ./client-run.sh 
Jan 06, 2015 5:03:14 PM com.nfsdb.journal.net.config.ClientConfig
INFO: Connected to localhost/127.0.0.1:7075
Client started

**Comment by bluestreak01** [comment]
thank you very much for trying these scenarios!

I managed to reproduce the problem. You can use "eth0" as ifName on both client and server, just increase receive buffer size on Client:

            setSoRcvBuf(10*1024*1024);

Here is my output:

Jan 07, 2015 12:03:49 AM com.nfsdb.net.mcast.AbstractOnDemandPoller
INFO: Polling on name:eth0 (eth0)
Jan 07, 2015 12:03:50 AM com.nfsdb.net.config.ClientConfig
WARNING: SO_SNDBUF value is ignored
Jan 07, 2015 12:03:50 AM com.nfsdb.net.config.ClientConfig
WARNING: SO_RCVBUF value is ignored
Jan 07, 2015 12:03:50 AM com.nfsdb.net.config.ClientConfig
INFO: Connected to /192.168.1.81:7075
Client started
Jan 07, 2015 12:03:50 AM com.nfsdb.net.StatsCollectingReadableByteChannel
INFO: received 117015697 bytes @ 306.579299 MB/s from: /192.168.1.81:7075 [352 calls]

(this is newer version but you should get similar results)

If you do get warnings - ignore them, receive buffer size is evidently not ignored.

Althogh replication between two computers works better with default buffer, on my WLAN it was done at 12MB/s. After increasing receive buffer size to 10MB replication speed went to 75MB/s. So it does make a difference for both cases.

Let me know if this works for you too.

**Comment by vguhesan** [comment]
Yes, that works! Thank you very much for helping me in debugging this. I would have never guessed at setting the "receive-buffer-size" parameter. 

Can you explain what made this work for my understanding?

In the code, I see that DEFAULT_SO_RCVBUF = 8192. In the code, I'm overriding the value to  10MB (1024*1024*10). Assumption is that the default value is represented as 8192MB. So how does going from 8MB to 10MB solve this problem? And also is there an optimal number to use based on my machine/nic interface?

On my Ubuntu machine here are the params for 
$ cat /proc/sys/net/core/rmem_default
212992
$ cat /proc/sys/net/core/rmem_max
212992

Is there a better value I must be using to maximize the bytes sent per payload?

Secondly, will I always need to do this on Linux or will future versions of NFSdb account for this?

Please advise.

I'm including my debug output just in case others run into this.

Thanks again for your help.

[server]
$ ./server-run.sh 
Jan 06, 2015 7:19:10 PM com.nfsdb.journal.net.config.ServerConfig
INFO: Server is now listening on /0.0.0.0:7075 [eth0]
Jan 06, 2015 7:19:10 PM com.nfsdb.journal.net.mcast.AbstractOnDemandSender
INFO: Sending on: name:eth0 (eth0)
Publishing: Jan 06, 2015 7:19:12 PM com.nfsdb.journal.net.JournalServer
INFO: Connected: /192.168.1.19:44373
.Jan 06, 2015 7:19:13 PM com.nfsdb.journal.net.StatsCollectingWritableByteChannel
INFO: sent 351000111 bytes @ 319.408197 MB/s to: /192.168.1.19:44373 [346 calls]
Jan 06, 2015 7:19:15 PM com.nfsdb.journal.net.StatsCollectingWritableByteChannel
INFO: sent 39000111 bytes @ 523.850763 MB/s to: /192.168.1.19:44373 [53 calls]
.Jan 06, 2015 7:19:16 PM com.nfsdb.journal.net.StatsCollectingWritableByteChannel
INFO: sent 39000111 bytes @ 39.992908 MB/s to: /192.168.1.19:44373 [52 calls]
.Jan 06, 2015 7:19:18 PM com.nfsdb.journal.net.StatsCollectingWritableByteChannel
INFO: sent 39000111 bytes @ 19.783726 MB/s to: /192.168.1.19:44373 [49 calls]
.Jan 06, 2015 7:19:20 PM com.nfsdb.journal.net.StatsCollectingWritableByteChannel
INFO: sent 39000111 bytes @ 17.898655 MB/s to: /192.168.1.19:44373 [53 calls]
.Jan 06, 2015 7:19:23 PM com.nfsdb.journal.net.StatsCollectingWritableByteChannel
INFO: sent 39000111 bytes @ 18.002616 MB/s to: /192.168.1.19:44373 [52 calls]
.Jan 06, 2015 7:19:25 PM com.nfsdb.journal.net.StatsCollectingWritableByteChannel
INFO: sent 39000111 bytes @ 17.941825 MB/s to: /192.168.1.19:44373 [49 calls]
.Jan 06, 2015 7:19:27 PM com.nfsdb.journal.net.StatsCollectingWritableByteChannel
INFO: sent 39000111 bytes @ 17.830012 MB/s to: /192.168.1.19:44373 [53 calls]
.Jan 06, 2015 7:19:29 PM com.nfsdb.journal.net.StatsCollectingWritableByteChannel
INFO: sent 39000111 bytes @ 17.643930 MB/s to: /192.168.1.19:44373 [48 calls]
.Jan 06, 2015 7:19:31 PM com.nfsdb.journal.net.StatsCollectingWritableByteChannel
INFO: sent 39000111 bytes @ 17.593853 MB/s to: /192.168.1.19:44373 [53 calls]
. [Done]

[client]
$ ./client-run.sh 
Jan 06, 2015 7:19:12 PM com.nfsdb.journal.net.mcast.AbstractOnDemandPoller
INFO: Polling on name:eth0 (eth0)
Jan 06, 2015 7:19:12 PM com.nfsdb.journal.net.config.ClientConfig
WARNING: SO_SNDBUF value is ignored
Jan 06, 2015 7:19:12 PM com.nfsdb.journal.net.config.ClientConfig
WARNING: SO_RCVBUF value is ignored
Jan 06, 2015 7:19:12 PM com.nfsdb.journal.net.config.ClientConfig
INFO: Connected to /192.168.1.19:7075
Client started
took: 8159345, count=9000000
Jan 06, 2015 7:19:15 PM com.nfsdb.journal.net.StatsCollectingReadableByteChannel
INFO: received 351000097 bytes @ 124.902902 MB/s from: /192.168.1.19:7075 [1818 calls]
took: 2488, count=1000000
Jan 06, 2015 7:19:15 PM com.nfsdb.journal.net.StatsCollectingReadableByteChannel
INFO: received 39000097 bytes @ 185.041746 MB/s from: /192.168.1.19:7075 [188 calls]
took: 1082, count=1000000
Jan 06, 2015 7:19:16 PM com.nfsdb.journal.net.StatsCollectingReadableByteChannel
INFO: received 39000097 bytes @ 98.656209 MB/s from: /192.168.1.19:7075 [193 calls]
took: 370, count=1000000
Jan 06, 2015 7:19:18 PM com.nfsdb.journal.net.StatsCollectingReadableByteChannel
INFO: received 39000097 bytes @ 199.964467 MB/s from: /192.168.1.19:7075 [194 calls]
took: 392, count=1000000
Jan 06, 2015 7:19:20 PM com.nfsdb.journal.net.StatsCollectingReadableByteChannel
INFO: received 39000097 bytes @ 180.550441 MB/s from: /192.168.1.19:7075 [202 calls]
took: 383, count=1000000
Jan 06, 2015 7:19:23 PM com.nfsdb.journal.net.StatsCollectingReadableByteChannel
INFO: received 39000097 bytes @ 199.964467 MB/s from: /192.168.1.19:7075 [194 calls]
took: 365, count=1000000
Jan 06, 2015 7:19:25 PM com.nfsdb.journal.net.StatsCollectingReadableByteChannel
INFO: received 39000097 bytes @ 206.629949 MB/s from: /192.168.1.19:7075 [188 calls]
took: 373, count=1000000
Jan 06, 2015 7:19:27 PM com.nfsdb.journal.net.StatsCollectingReadableByteChannel
INFO: received 39000097 bytes @ 206.629949 MB/s from: /192.168.1.19:7075 [210 calls]
took: 410, count=1000000
Jan 06, 2015 7:19:29 PM com.nfsdb.journal.net.StatsCollectingReadableByteChannel
INFO: received 39000097 bytes @ 187.845408 MB/s from: /192.168.1.19:7075 [193 calls]
took: 436, count=1000000
Jan 06, 2015 7:19:31 PM com.nfsdb.journal.net.StatsCollectingReadableByteChannel
INFO: received 39000097 bytes @ 182.320543 MB/s from: /192.168.1.19:7075 [199 calls]

**Comment by bluestreak01** [comment]
I'm pleased we found a solution.

To be honest it was a bit of a hunch. Bmon was showing that despite setting ifName to "eth0" traffic was going over localhost and it was going at a very slow rate, 50b/s. I know for a fact that i'm writing large buffers to SocketChannel, something like 16MB buffer at a time. There could be two reasons for slow transfer speed, either packet loss or tiny transfer buffer. "lo" wouldn't be losing packets so I tried buffer.

1MB works just as well as 10MB. I'm on windows soRcvBuf makes no difference. I'll try to find optimal value in a few mins, need to flick back to ubuntu :)
